File: android/app/src/main/java/com/stak/demo/ui/components/RangeChart.kt
package com.stak.demo.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.dp

/** The six range labels every chart offers; "3M" is the authored default and keeps each frame's exported line. */
internal val RANGE_LABELS = listOf("1D", "1W", "1M", "3M", "YTD", "1Y")

/**
 * Demo stand-ins until the backend serves price history - each range's line
 * as fractions of the chart height from the bottom (0 = bottom), spread
 * evenly across the width. "3M" is absent on purpose: that range shows the
 * authored export (sim_chart_line 1:3935, ms_chart_line 1:3155, sd_chart_line
 * 1:2382). Shared by Simulate, My STAK, Stock Detail and Pick Detail (user,
 * 2026-09-05: "be able to click on the timeline"). Mirrors
 * ios/StakDemo/Theme/RangeChart.swift.
 */
internal val RANGE_SERIES: Map<String, List<Float>> = mapOf(
	"1D" to listOf(0.45f, 0.50f, 0.42f, 0.55f, 0.60f, 0.52f, 0.58f, 0.66f, 0.62f, 0.70f),
	"1W" to listOf(0.30f, 0.38f, 0.35f, 0.50f, 0.46f, 0.60f, 0.72f),
	"1M" to listOf(0.25f, 0.30f, 0.28f, 0.42f, 0.38f, 0.52f, 0.48f, 0.60f, 0.55f, 0.68f, 0.75f),
	"YTD" to listOf(0.20f, 0.35f, 0.30f, 0.45f, 0.40f, 0.55f, 0.50f, 0.62f, 0.70f, 0.66f, 0.80f),
	"1Y" to listOf(0.15f, 0.22f, 0.30f, 0.26f, 0.40f, 0.36f, 0.50f, 0.55f, 0.48f, 0.62f, 0.70f, 0.82f),
)

/**
 * A range's line: a `tint` 2-wide round stroke in the caller's box and
 * nothing else - every authored "Chart line" (1:3245, 1:3936, 1:2409,
 * 1:4664) is a bare #69B3CA 2 stroke with no fill (user, 2026-09-05: "was
 * there a gradient?" - there was not; the old fill came from a stand-in).
 */
@Composable
internal fun RangeChart(series: List<Float>, tint: Color, modifier: Modifier) {
	val u = com.stak.demo.ui.onboarding.figmaUnit()
	Canvas(modifier = modifier) {
		val w = size.width
		val h = size.height
		val last = series.lastIndex
		val points = series.mapIndexed { i, f -> Offset(if (last == 0) 0f else w * i / last, h * (1f - f)) }
		val line = Path().apply {
			moveTo(points.first().x, points.first().y)
			points.drop(1).forEach { lineTo(it.x, it.y) }
		}
		drawPath(line, tint, style = Stroke(width = (2 * u).dp.toPx(), cap = StrokeCap.Round, join = StrokeJoin.Round))
	}
}

File: ios/StakDemo/Theme/RangeChart.swift
import SwiftUI

/// The six range labels every chart offers; "3M" is the authored default and keeps each frame's exported line.
let rangeLabels = ["1D", "1W", "1M", "3M", "YTD", "1Y"]

/// A chart at a range. "3M" shows the authored export (`authored`); every
/// other range draws the shared demo series in the same box - a `tint`
/// 2-wide round stroke and nothing else: every authored "Chart line" is a
/// bare #69B3CA 2 stroke with no fill (user, 2026-09-05). The series are demo
/// stand-ins until the backend serves price history. Shared by Simulate,
/// My STAK, Stock Detail and Pick Detail (user, 2026-09-05: "be able to
/// click on the timeline"). Mirrors android ui/components/RangeChart.kt.
struct RangeLineChart: View {
	let range: String
	let tint: Color
	let authored: String
	let width: CGFloat
	let height: CGFloat

	/// Height fraction from the bottom (0 = bottom) per point, spread
	/// evenly across the width.
	static let series: [String: [CGFloat]] = [
		"1D": [0.45, 0.50, 0.42, 0.55, 0.60, 0.52, 0.58, 0.66, 0.62, 0.70],
		"1W": [0.30, 0.38, 0.35, 0.50, 0.46, 0.60, 0.72],
		"1M": [0.25, 0.30, 0.28, 0.42, 0.38, 0.52, 0.48, 0.60, 0.55, 0.68, 0.75],
		"YTD": [0.20, 0.35, 0.30, 0.45, 0.40, 0.55, 0.50, 0.62, 0.70, 0.66, 0.80],
		"1Y": [0.15, 0.22, 0.30, 0.26, 0.40, 0.36, 0.50, 0.55, 0.48, 0.62, 0.70, 0.82]
	]

	var body: some View {
		let u = figmaUnit
		Group {
			if let points = RangeLineChart.series[range], range != "3M" {
				GeometryReader { geo in
					let line = RangeLineChart.linePath(points, in: geo.size)
					line.stroke(tint, style: StrokeStyle(lineWidth: 2 * u, lineCap: .round, lineJoin: .round))
				}
			} else {
				Image(authored)
					.resizable()
					.scaledToFit()
			}
		}
		.frame(width: width, height: height)
	}

	/// The series as a polyline across the box.
	private static func linePath(_ points: [CGFloat], in size: CGSize) -> Path {
		var path = Path()
		let steps = CGFloat(max(points.count - 1, 1))
		for (i, fraction) in points.enumerated() {
			let point = CGPoint(x: size.width * CGFloat(i) / steps, y: size.height * (1 - fraction))
			if i == 0 { path.move(to: point) } else { path.addLine(to: point) }
		}
		return path
	}
}

/// The authored 39x22.5 teal pill row; `selected` follows the tap.
struct RangePills: View {
	@Binding var selected: String
	let tint: Color
	let muted: Color

	var body: some View {
		let u = figmaUnit
		HStack(spacing: 37 * u) {
			ForEach(rangeLabels, id: \.self) { label in
				Button { selected = label } label: {
					if label == selected {
						Text(label)
							.font(StakFont.geist(12 * u, .medium))
							.foregroundStyle(tint)
							.frame(width: 39 * u, height: 22.5 * u)
							.background(Color(argb: 0x292C9DBC), in: RoundedRectangle(cornerRadius: 11.25 * u))
							.overlay(
								RoundedRectangle(cornerRadius: 11.25 * u)
									.strokeBorder(Color(argb: 0x662C9DBC), lineWidth: 0.75 * u)
							)
					} else {
						Text(label)
							.font(StakFont.geist(12 * u))
							.foregroundStyle(muted)
					}
				}
				.buttonStyle(.plain)
			}
		}
	}
}
